flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSumAggFunction.java

```diff
@@ -0,0 +1,167 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.functions.hive;
+
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.expressions.Expression;
+import org.apache.flink.table.expressions.UnresolvedCallExpression;
+import org.apache.flink.table.expressions.UnresolvedReferenceExpression;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.inference.CallContext;
+
+import static org.apache.flink.table.expressions.ApiExpressionUtils.unresolvedRef;
+import static org.apache.flink.table.functions.BuiltInFunctionDefinitions.UNIX_TIMESTAMP;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.call;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.hiveAggDecimalPlus;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.ifThenElse;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.isNull;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.nullOf;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.plus;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.tryCast;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.typeLiteral;
+import static org.apache.flink.table.types.logical.DecimalType.MAX_PRECISION;
+import static org.apache.flink.table.types.logical.LogicalTypeRoot.DECIMAL;
+import static org.apache.flink.table.types.logical.LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getPrecision;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getScale;
+
+/** built-in hive sum aggregate function. */
+public class HiveSumAggFunction extends HiveDeclarativeAggregateFunction {
+
+    private final UnresolvedReferenceExpression sum = unresolvedRef("sum");
+    private DataType argsType;
+    private DataType resultType;
+
+    @Override
+    public int operandCount() {
+        return 1;
+    }
+
+    @Override
+    public UnresolvedReferenceExpression[] aggBufferAttributes() {
+        return new UnresolvedReferenceExpression[] {sum};
+    }
+
+    @Override
+    public DataType[] getAggBufferTypes() {
+        return new DataType[] {getResultType()};
+    }
+
+    @Override
+    public DataType getResultType() {
+        return resultType;
+    }
+
+    @Override
+    public Expression[] initialValuesExpressions() {
+        return new Expression[] {/* sum = */ nullOf(getResultType())};
+    }
+
+    @Override
+    public Expression[] accumulateExpressions() {
+        Expression operand;
+        // TimestampToNumericCastRule can't cast timestamp to numeric directly, so here use
+        // UNIX_TIMESTAMP(CAST(timestamp_col AS STRING)) instead
+        if (argsType.getLogicalType().is(TIMESTAMP_WITHOUT_TIME_ZONE)) {
```

Comment:
Does hive sum udaf even support input parameter of `TIMESTAMP_WITHOUT_TIME_ZONE` type?

Reply:
Yes，[sum](https://github.com/apache/hive/blob/10805bc997d7cd136b85fca9200cf165ffe2eae5/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFSum.java#L123) udaf supports.

Reply:
BTW, the hive [TimestampObjectInspector](https://github.com/apache/hive/blob/989e72a393356c5f91f96d1bab6455a4c75c77a7/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/primitive/PrimitiveObjectInspectorUtils.java#L874) support return double precision, but flink  `UNIX_TIMESTAMP ` only can return bigint, so there may be a loss of precision here, for example `2021-08-04 16:26:33.4`, flink only can get the part of `2021-08-04 16:26:33`, the `.4` is loss. For timestamp type, we cannot align the behavior with hive, what do you think about it?

Reply:
@lsyldliu Thanks for pointing the problem out.
IMO, it's ok to not support summing over timestamp types.
Intuitively, sum(timestamp) does not seem to make sense. Besides the behavior of flink hive dialect is not easy to be consistent with hive behavior.
WDYT?

Reply:
I agree with you.